vortex-array/src/arrays/primitive/top_value.rs

```diff
@@ -0,0 +1,58 @@
+use std::hash::Hash;
+
+use rustc_hash::FxBuildHasher;
+use vortex_dtype::{NativePType, match_each_native_ptype};
+use vortex_error::{VortexExpect, VortexResult, vortex_bail};
+use vortex_mask::{AllOr, Mask};
+use vortex_scalar::PValue;
+
+use crate::Array;
+use crate::aliases::hash_map::HashMap;
+use crate::arrays::{NativeValue, PrimitiveArray};
+use crate::variants::PrimitiveArrayTrait;
+
+impl PrimitiveArray {
+    pub fn top_value(&self) -> VortexResult<(PValue, u32)> {
+        if self.is_empty() {
+            vortex_bail!("Can't compute top value for empty array")
+        }
+
+        if self.all_invalid()? {
+            vortex_bail!("Can't compute top value for all null array")
+        }
+
+        match_each_native_ptype!(self.ptype(), |$P| {
+            typed_top_value(self.as_slice::<$P>(), self.validity_mask()?).map(|(v, c)|  (v.into(), c))
+        })
+    }
+}
+
+fn typed_top_value<T>(values: &[T], mask: Mask) -> VortexResult<(T, u32)>
+where
+    T: NativePType,
+    NativeValue<T>: Eq + Hash,
+{
+    let mut distinct_values: HashMap<NativeValue<T>, u32, FxBuildHasher> =
+        HashMap::with_hasher(FxBuildHasher);
+    match mask.boolean_buffer() {
+        AllOr::All => {
+            for value in values.iter().copied() {
+                *distinct_values.entry(NativeValue(value)).or_insert(0) += 1;
```

Comment:
this is kind of dumb, but if we have an array with more than u32::MAX elements, this will overflow in release, when we want it to saturate. should just use `saturating_add`

Reply:
we can just make it usize